src/imp/unix.rs

```diff
@@ -99,24 +114,30 @@ pub fn reopen(file: &File, path: &Path) -> io::Result<File> {
     }
 }
 
+#[cfg(not(target_os = "redox"))]
 pub fn persist(old_path: &Path, new_path: &Path, overwrite: bool) -> io::Result<()> {
     unsafe {
         let old_path = try!(cstr(old_path));
         let new_path = try!(cstr(new_path));
         if overwrite {
-            if rename(old_path.as_ptr() as *const c_char,
-                      new_path.as_ptr() as *const c_char) != 0 {
-                return Err(io::Error::last_os_error());
-            }
+            try!(cvt_err(rename(old_path.as_ptr() as *const c_char,
+                                new_path.as_ptr() as *const c_char)));
         } else {
-            if link(old_path.as_ptr() as *const c_char,
-                    new_path.as_ptr() as *const c_char) != 0 {
-                return Err(io::Error::last_os_error());
-            }
+            try!(cvt_err(link(old_path.as_ptr() as *const c_char,
+                              new_path.as_ptr() as *const c_char)));
             // Ignore unlink errors. Can we do better?
             // On recent linux, we can use renameat2 to do this atomically.
             let _ = unlink(old_path.as_ptr() as *const c_char);
         }
         Ok(())
     }
 }
+
+#[cfg(target_os = "redox")]
+pub fn persist(old_path: &Path, new_path: &Path, overwrite: bool) -> io::Result<()> {
+    // XXX implement in better way when possible
+    if !overwrite && new_path.exists() {
```

Comment:
Wait... by "copy and unlink" you mean it literally *copies* the file byte-by-byte? If that's the case, I think the only sane implementation here would be to return an error. The entire point of this method is that it allows the programmer to atomically replace an existing file with an updated file (and I've seen users rely on this guarantee).